### PR TITLE
OCPBUGS#9215: Updated port 1936 in install vSphere docs

### PR DESCRIPTION
**Pull request by dfitzmau**
[OCPBUGS-9215](https://issues.redhat.com/browse/OCPBUGS-9215)

Version(s):
4.14 through to 4.10

Link to docs preview:
[Load balancing requirements for user-provisioned infrastructure](https://61430--docspreview.netlify.app/openshift-enterprise/latest/installing/installing_vsphere/installing-vsphere.html#installation-load-balancing-user-infra_installing-vsphere)

QE review:
- [ ] QE has approved this change.

Additional information:
* [KCS](https://access.redhat.com/solutions/6964370)

### CODE REVIEW

**Comment by palonsoro** [comment]
> If you need to use port 1936, ensure that you configure  this port through your firewall’s configuration, because each control  plane node needs to access this port. Do not use the Ingress load balancer to expose this port.

Not sure why it is mentioned "If you need to use port 1936". Users never need this port (unless they want to install a custom monitoring) but control plane does. So it must be accessible within the cluster (regardless of whether the user does something with it or not) but not included in the load balancer (as correctly indicated in the latest sentence)

**Comment by palonsoro** [comment]
> You must configure the Ingress router after the control plane initializes.

Not sure on the reason for this sentence. The ingress load balancer (including the layer 4 condition described before) must be configured before starting the installation. It is true that ingress controller customization has to be done as 2-day operation, but bringing it here in this sentence may be confusing (because customer may misinterpret that the load balancer has to be configured too late).

**Comment by palonsoro** [comment]
And the most important failure: This change is not removing the 1936 port from the load balancer example. We need the customers to not expose the 1936 port in the load balancer (the very reason for the original bug report), so we also should not make our own example expose it.

**Comment by dfitzmau** [comment]
> If you need to use port

Thanks, @palonsoro . 

> > If you need to use port 1936, ensure that you configure  this port through your firewall’s configuration, because each control  plane node needs to access this port. Do not use the Ingress load balancer to expose this port.
> 
> Not sure why it is mentioned "If you need to use port 1936". Users never need this port (unless they want to install a custom monitoring) but control plane does. So it must be accessible within the cluster (regardless of whether the user does something with it or not) but not included in the load balancer (as correctly indicated in the latest sentence)

Thanks, @palonsoro .

How about:

> By default, port `1936` is accessible for a {product-title} cluster, because each control plane node needs access to this port. Do not use the Ingress load balancer to expose this port.


What would happen if a user did expose the port with the load balancer?

**Comment by dfitzmau** [comment]
> And the most important failure: This change is not removing the 1936 port from the load balancer example. We need the customers to not expose the 1936 port in the load balancer (the very reason for the original bug report), so we also should not make our own example expose it.

Thanks, @palonsoro . This might be confusing because of the following sentence:

> Configure the following ports on both the front and back of the load balancers

**Comment by dfitzmau** [comment]
> must configure the Ingress router after the control plan

Thanks, @palonsoro . Remove the states sentence and the `Layer 4 load balancing only` bullet point?

Screenshot from 2023-06-20 11-37-20

**Comment by palonsoro** [comment]
> What would happen if a user did expose the port with the load balancer?

They would hit the stats of one of the router pods, chosen at random, which makes no sense at all. This is why we should not recommend exposing them, even slightly.

**Comment by palonsoro** [comment]
> Unless "You must configure the /healthz/ready endpoint for the Ingress health check probe." is a different type of configuration that is separate from exposing the LB?

I don't understand what you mean here.

The table 11 indicating which ports have to be exposed is correct. But the example is also exposing the 1936 port, which is not correct. It must not expose it. And here, "expose" means "have a frontend that listens on it and forwards to a backend with that port". Shorter explained: It must just not be configured at all in the front load balancer.

The meaning of the older (and now removed) statement of configuring health probes for the load balancer is too vague (which was a mistake on its own).  Trying to guess, it might refer to either of the following:
- If you configured the 1936 port, you had to set that path for the health check. As we are removing it, we don't care.
- They might have been suggesting to use the `/healthz/ready` path of port 1936 for the health checks of the main 80 and 443 ports. That would be feasible but by making the example much more complex than it is (and I am still not fully convinced that it would be an improvement, because it wouldn't detect a firewall issue, for example). I have been searching the cluster-ingress-operator code to see if they tried something like this while creating the load balancers and they didn't, so we shouldn't try this convoluted approach in what should be a simple example.

In short, let's just remove the references to the 1936 port from the example :-) 

So, in short, we should keep things simple and just remove the references

**Comment by palonsoro** [comment]
> Thanks, @palonsoro . Remove the states sentence and the Layer 4 load balancing only bullet point?

I'd just remove the sentence that says `You must configure the Ingress router after the control plane initializes.` sentence and I wouldn't touch anything else (except maybe some rewriting without meaning change if the text around the removed sentence requires it to make sense).

**Comment by dfitzmau** [comment]
> By default, port `1936` is accessible for a {product-title} cluster, because each control plane node needs access to this port. Do not use the Ingress load balancer to expose this port.

Thanks, updated to:

> By default, port `1936` is accessible for a {product-title} cluster, because each control plane node needs access to this port. Do not use the Ingress load balancer to expose this port, because exposing the port in this way would display random statistics from the Ingress router pod and would expose internal router components externally to the cluster.




> > Unless "You must configure the /healthz/ready endpoint for the Ingress health check probe." is a different type of configuration that is separate from exposing the LB?
> 
> I don't understand what you mean here.
> 
> The table 11 indicating which ports have to be exposed is correct. But the example is also exposing the 1936 port, which is not correct. It must not expose it. And here, "expose" means "have a frontend that listens on it and forwards to a backend with that port". Shorter explained: It must just not be configured at all in the front load balancer.
> 
> The meaning of the older (and now removed) statement of configuring health probes for the load balancer is too vague (which was a mistake on its own). Trying to guess, it might refer to either of the following:
> 
> * If you configured the 1936 port, you had to set that path for the health check. As we are removing it, we don't care.
> * They might have been suggesting to use the `/healthz/ready` path of port 1936 for the health checks of the main 80 and 443 ports. That would be feasible but by making the example much more complex than it is (and I am still not fully convinced that it would be an improvement, because it wouldn't detect a firewall issue, for example). I have been searching the cluster-ingress-operator code to see if they tried something like this while creating the load balancers and they didn't, so we shouldn't try this convoluted approach in what should be a simple example.
> 
> In short, let's just remove the references to the 1936 port from the example :-)
> 
> So, in short, we should keep things simple and just remove the references

Thanks, @palonsoro . Including the`1936` reference in the `.cfg` configuration file example in `Example load balancer configuration for user-provisioned clusters`?

**Comment by sgaoshang** [comment]
/lgtm, cc: @lihongan Do you have concerns with removing it?

1936 | The worker nodes that run the Ingress Controller pods, by default. You must configure the /healthz/ready endpoint for the ingress health check probe. | X | X | HTTP traffic

**Comment by palonsoro** [comment]
> Including the1936 reference in the .cfg configuration file example in Example load balancer configuration for user-provisioned clusters?

Not sure to understand this question. But we need to fully remove all references from the haproxy configuration example. I can help in doing so if you need.

**Comment by palonsoro** [comment]
Ok. I made a mistake: re-reading the haproxy configuration, I see that the stats exposed are the ones of the front haproxy load balancer itself, not the ones from the routers.

I'd still recommend removing them, because that's the first thing that most security scanners would recommend (and they are of no value in what regards exemplifying an haproxy configuration for OpenShift). 

It would mean removing this:
```
frontend stats
  bind *:1936 <1>
  mode            http
  log             global
  maxconn 10
  stats enable
  stats hide-version
  stats refresh 30s
  stats show-node
  stats show-desc Stats for ocp4 cluster <2>
  stats auth admin:ocp4
  stats uri /stats
```

And the `<1>` and `<2>` bullet points present on it.

However, at least, it is not as much of a security risk as I thought (yet still safer if removed).

**Comment by dfitzmau** [comment]
> Ok. I made a mistake: re-reading the haproxy configuration, I see that the stats exposed are the ones of the front haproxy load balancer itself, not the ones from the routers.
> 
> I'd still recommend removing them, because that's the first thing that most security scanners would recommend (and they are of no value in what regards exemplifying an haproxy configuration for OpenShift).
> 
> It would mean removing this:
> 
> ```
> frontend stats
>   bind *:1936 <1>
>   mode            http
>   log             global
>   maxconn 10
>   stats enable
>   stats hide-version
>   stats refresh 30s
>   stats show-node
>   stats show-desc Stats for ocp4 cluster <2>
>   stats auth admin:ocp4
>   stats uri /stats
> ```
> 
> And the `<1>` and `<2>` bullet points present on it.
> 
> However, at least, it is not as much of a security risk as I thought (yet still safer if removed).

Thanks, Pablo. I removed this the `1936` configuration for this HAProxy example.

**Comment by palonsoro** [comment]
Perfect!

Thx!

/lgtm

**Comment by dfitzmau** [comment]
Just waiting for a confirmation from @lihongan on the changes.

**Comment by lihongan** [comment]
LGTM. Thank you @dfitzmau

File: modules/installation-infrastructure-user-infra.adoc [hunk 1]
```diff
@@ -74,6 +74,13 @@ endif::ibm-z-kvm[]
 . Ensure that your network infrastructure provides the required network connectivity between the cluster components. See the _Networking requirements for user-provisioned infrastructure_ section for details about the requirements.
 
 . Configure your firewall to enable the ports required for the {product-title} cluster components to communicate. See _Networking requirements for user-provisioned infrastructure_ section for details about the ports that are required.
++
+[IMPORTANT]
+====
+By default, port `1936` is accessible for a {product-title} cluster, because each control plane node needs access to this port.
```
Comment (snarayan-redhat) [inline modules/installation-infrastructure-user-infra.adoc:80]:
```suggestion
By default, the port `1936` is accessible for an {product-title} cluster, because each control plane node needs access to this port. 
```

File: modules/installation-infrastructure-user-infra.adoc [hunk 2]
```diff
@@ -74,6 +74,13 @@ endif::ibm-z-kvm[]
 . Ensure that your network infrastructure provides the required network connectivity between the cluster components. See the _Networking requirements for user-provisioned infrastructure_ section for details about the requirements.
 
 . Configure your firewall to enable the ports required for the {product-title} cluster components to communicate. See _Networking requirements for user-provisioned infrastructure_ section for details about the ports that are required.
++
+[IMPORTANT]
+====
+By default, port `1936` is accessible for a {product-title} cluster, because each control plane node needs access to this port. 
+
+Do not use the Ingress load balancer to expose this port, because exposing the port in this way could expose sensitive information, such as statistics and metrics, related to Ingress Controllers.
```
Comment (snarayan-redhat) [inline modules/installation-infrastructure-user-infra.adoc:82]:
```suggestion
Avoid using the Ingress load balancer to expose this port, because doing so might result in the exposure of sensitive information, such as statistics and metrics, related to Ingress Controllers.
```

File: modules/installation-load-balancing-user-infra.adoc [hunk 3]
```diff
@@ -169,45 +171,34 @@ Configure the following ports on both the front and back of the load balancers:
...
 == Example load balancer configuration for clusters that are deployed with user-managed load balancers
 
-This section provides an example API and application ingress load balancer configuration that meets the load balancing requirements for clusters that are deployed with user-managed load balancers. The sample is an `/etc/haproxy/haproxy.cfg` configuration for an HAProxy load balancer. The example is not meant to provide advice for choosing one load balancing solution over another.
+This section provides an example API and application Ingress load balancer configuration that meets the load balancing requirements for clusters that are deployed with user-managed load balancers. The sample is an `/etc/haproxy/haproxy.cfg` configuration for an HAProxy load balancer. The example is not meant to provide advice for choosing one load balancing solution over another.
 endif::user-managed-lb[]
 
+In the example, the same load balancer is used for the Kubernetes API and application ingress traffic. In production scenarios you can deploy the API and application ingress load balancers separately so that you can scale the load balancer infrastructure for each in isolation.
```
Comment (snarayan-redhat) [inline modules/installation-load-balancing-user-infra.adoc:194]:
```suggestion
In the example, the same load balancer is used for the Kubernetes API and application ingress traffic. In production scenarios, you can deploy the API and application ingress load balancers separately so that you can scale the load balancer infrastructure for each in isolation.
```

File: modules/installation-network-user-infra.adoc [hunk 4]
```diff
@@ -193,7 +193,7 @@ endif::ibm-z-kvm[]
 
 .4+|TCP
 |`1936`
-|Metrics
+|Metrics. Can receive internal and external HTTPS traffic.
```
Comment (snarayan-redhat) [inline modules/installation-network-user-infra.adoc:196]:
```suggestion
|Metrics. Can receive the internal and external HTTPS traffic.
```

**Review by snarayan-redhat — COMMENTED** [review]
Minor nits and suggestions. Overall looks good!

File: modules/installation-network-user-infra.adoc (hunk 4 above)
Comment (palonsoro) [reply modules/installation-network-user-infra.adoc:196]: This suggestion would be wrong, in the sense that this is not the only port which receives HTTPS traffic (it is not even the main one), so saying that it may receive **the** internal and external HTTPS traffic, can lead to the false impression that it is the one and only HTTPS port and be extremely misleading.

Comment (dfitzmau) [reply modules/installation-network-user-infra.adoc:196]: Hi @palonsoro . Thanks for the comment. Would stating "both internal and external HTTPS traffic" work for you?

Comment (dfitzmau) [reply modules/installation-network-user-infra.adoc]: Thanks, @palonsoro . I've removed this sentence from the table entry.

File: modules/installation-network-user-infra.adoc [hunk 5]
```diff
@@ -193,7 +193,7 @@ endif::ibm-z-kvm[]
 
 .4+|TCP
 |`1936`
-|Metrics
+|Metrics.
```
Comment (stevsmit) [inline modules/installation-network-user-infra.adoc:196]: I'm not sure if I'd add a period to this or not. The only other ones in the table that have periods are complete sentences. If we keep the period we might want to add a period for the following, and maybe others (seems like a can of worms):

**Network reachability tests**

**The default ports that Kubernetes reserves**

**Kubernetes node port**

Comment (dfitzmau) [reply modules/installation-network-user-infra.adoc]: Thanks, @stevsmit . Valid point. I removed the period.

**Comment by stevsmit** [comment]
LGTM
